File: apps/web/src/api/harness.ts
import { apiClient } from './client';

// ── Harness Types ─────────────────────────────────────────────────────────────

export type HarnessWorkflowType =
  | 'signal_extraction'
  | 'account_intelligence'
  | 'value_model_generation'
  | 'evidence_matching'
  | 'value_tree_generation'
  | 'roi_calculator_generation'
  | 'business_case_generation'
  | 'renewal_risk_analysis'
  | 'expansion_opportunity_analysis';

export type HarnessRunStatus =
  | 'queued'
  | 'running'
  | 'waiting_for_human'
  | 'failed'
  | 'completed'
  | 'cancelled';

export type HarnessState =
  | 'INIT'
  | 'RESOLVE_CONTEXT'
  | 'LOAD_VALUE_PACK'
  | 'RETRIEVE_KNOWLEDGE'
  | 'GENERATE_HYPOTHESES'
  | 'MATCH_EVIDENCE'
  | 'QUANTIFY_IMPACT'
  | 'VALIDATE_CLAIMS'
  | 'HUMAN_REVIEW'
  | 'PUBLISH_OUTPUT'
  | 'DONE'
  | 'FAILED'
  | 'CANCELLED';

export type GateType =
  | 'approve_claims'
  | 'approve_assumptions'
  | 'approve_customer_output'
  | 'resolve_conflict';

export type GateStatus = 'pending' | 'approved' | 'rejected' | 'modified' | 'expired';

export type GateDecision = 'approved' | 'rejected' | 'modified' | 'expired';

export type ValidationState =
  | 'passed'
  | 'failed'
  | 'needs_review'
  | 'insufficient_evidence';

export type InitiatedBy = 'user' | 'system' | 'agent' | 'scheduled_job';

export const TERMINAL_STATES: HarnessState[] = ['DONE', 'FAILED', 'CANCELLED'];

export function isTerminalState(state: HarnessState): boolean {
  return TERMINAL_STATES.includes(state);
}

// ── Domain Models ─────────────────────────────────────────────────────────────

export interface HarnessRun {
  id: string;
  tenant_id: string;
  account_id: string | null;
  workflow_type: HarnessWorkflowType;
  initiated_by: InitiatedBy;
  status: HarnessRunStatus;
  current_state: HarnessState;
  value_pack_id: string | null;
  trace_id: string;
  created_at: string;
  updated_at: string;
}

export interface HarnessCheckpoint {
  id: string;
  run_id: string;
  tenant_id: string;
  state_name: HarnessState;
  input_hash: string;
  output_hash: string | null;
  created_at: string;
}

export interface HumanGate {
  id: string;
  run_id: string;
  tenant_id: string;
  gate_type: GateType;
  status: GateStatus;
  decision_by: string | null;
  decision_reason: string | null;
  created_at: string;
  decided_at: string | null;
}

export interface ClaimValidationResult {
  id: string;
  tenant_id: string;
  claim_id: string;
  validation_state: ValidationState;
  evidence_refs: string[];
  confidence: number;
  trust_score: number;
  validator: string;
  reason: string;
  created_at: string;
}

// ── Request / Response Types ──────────────────────────────────────────────────

export interface CreateRunRequest {
  workflow_type: HarnessWorkflowType;
  initiated_by?: InitiatedBy;
  account_id?: string;
  value_pack_id?: string;
}

export interface RunListResponse {
  items: HarnessRun[];
  total: number;
  limit: number;
  offset: number;
  has_more: boolean;
}

export interface TransitionRequest {
  to_state: HarnessState;
  human_override?: boolean;
  state_payload?: Record<string, unknown>;
  validation_results?: ClaimValidationResult[];
}

export interface TransitionResponse {
  run: HarnessRun;
  trace_event: {
    trace_id: string;
    run_id: string;
    from_state: HarnessState | null;
    to_state: HarnessState | null;
    event_type: string;
    timestamp: string;
  } | null;
}

export interface CheckpointListResponse {
  items: HarnessCheckpoint[];
  total: number;
}

export interface GateListResponse {
  items: HumanGate[];
  total: number;
}

export interface GateDecisionRequest {
  decision: GateDecision;
  decision_reason?: string;
}

export interface ClaimToValidate {
  claim_id: string;
  claim_text: string;
  evidence_refs?: string[];
  value_pack_id?: string;
  account_id?: string;
}

export interface ValidateClaimsRequest {
  claims: ClaimToValidate[];
}

export interface ValidateClaimsResponse {
  results: ClaimValidationResult[];
  total: number;
  passed: number;
  failed: number;
  needs_review: number;
  insufficient_evidence: number;
}

export interface HarnessHealthResponse {
  status: 'ok' | 'degraded';
  validation_available: boolean;
  l5_healthy: boolean;
  db_healthy: boolean;
}

// ── API Client ────────────────────────────────────────────────────────────────

const L4 = 'l4' as const;

export const harnessApi = {
  // Runs
  createRun: (data: CreateRunRequest) =>
    apiClient.post(L4, '/harness/runs', data).then(r => r.data as HarnessRun),

  listRuns: (params?: {
    status?: HarnessRunStatus;
    workflow_type?: HarnessWorkflowType;
    limit?: number;
    offset?: number;
  }) => {
    const query = new URLSearchParams();
    if (params?.status) query.set('status', params.status);
    if (params?.workflow_type) query.set('workflow_type', params.workflow_type);
    if (params?.limit != null) query.set('limit', String(params.limit));
    if (params?.offset != null) query.set('offset', String(params.offset));
    const qs = query.toString();
    return apiClient.get(L4, `/harness/runs${qs ? `?${qs}` : ''}`).then(r => r.data as RunListResponse);
  },

  getRun: (runId: string) =>
    apiClient.get(L4, `/harness/runs/${runId}`).then(r => r.data as HarnessRun),

  transitionRun: (runId: string, data: TransitionRequest) =>
    apiClient.post(L4, `/harness/runs/${runId}/transition`, data).then(r => r.data as TransitionResponse),

  cancelRun: (runId: string) =>
    apiClient.delete(L4, `/harness/runs/${runId}`).then(r => r.data),

  // Checkpoints
  listCheckpoints: (runId: string) =>
    apiClient.get(L4, `/harness/runs/${runId}/checkpoints`).then(r => r.data as CheckpointListResponse),

  getLatestCheckpoint: (runId: string) =>
    apiClient.get(L4, `/harness/runs/${runId}/checkpoints/latest`).then(r => r.data as HarnessCheckpoint),

  // Gates
  listGates: (runId: string) =>
    apiClient.get(L4, `/harness/runs/${runId}/gates`).then(r => r.data as GateListResponse),

  createGate: (runId: string, data: { gate_type: GateType }) =>
    apiClient.post(L4, `/harness/runs/${runId}/gates`, data).then(r => r.data as HumanGate),

  decideGate: (gateId: string, data: GateDecisionRequest) =>
    apiClient.post(L4, `/harness/gates/${gateId}/decide`, data).then(r => r.data as HumanGate),

  // Validation
  validateClaims: (runId: string, data: ValidateClaimsRequest) =>
    apiClient.post(L4, `/harness/runs/${runId}/validate`, data).then(r => r.data as ValidateClaimsResponse),

  // Health
  health: () =>
    apiClient.get(L4, '/harness/health').then(r => r.data as HarnessHealthResponse),
};

File: services/layer4-agents/src/api/routes/harness.py
"""FastAPI routes for the Fabric Harness.

Exposes the SqlHarnessRegistry as a REST API under /v1/harness/*.

All routes:
  - Extract tenant_id from authenticated RequestContext (never from request body).
  - Inject SqlHarnessRegistry via get_harness_registry() dependency.
  - Return 404 on missing resources, 403 on tenant mismatch, 400 on bad input.
  - Follow the same error shape as /v1/workflows/*.
"""

from __future__ import annotations

import logging
import os
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from value_fabric.shared.identity.context import RequestContext
from value_fabric.shared.identity.dependencies import require_authenticated, require_content_admin

from ...database import get_db_from_context
from ...harness.api_models import (
    CheckpointListResponse,
    CheckpointResponse,
    CreateGateRequest,
    CreateRunRequest,
    GateDecisionRequest,
    GateListResponse,
    GateResponse,
    HarnessHealthResponse,
    RunListResponse,
    RunResponse,
    TraceEventResponse,
    TransitionRequest,
    TransitionResponse,
    ValidateClaimsRequest,
    ValidateClaimsResponse,
    ValidationResultResponse,
)
from ...harness.factory import make_live_l5_registry, make_sql_registry
from ...harness.models import GateStatus, HarnessRunStatus, HarnessState, HarnessWorkflowType
from ...harness.registry import HarnessRegistryError
from ...harness.sql_stores import SqlHarnessRegistry
from ...harness.validation_hooks import ClaimValidationRequest as DomainClaimValidationRequest

logger = logging.getLogger(__name__)

router = APIRouter(tags=["harness"])


# ---------------------------------------------------------------------------
# Dependency: registry
# ---------------------------------------------------------------------------


async def get_harness_registry(
    db: AsyncSession = Depends(get_db_from_context),
    _ctx: RequestContext = Depends(require_authenticated),
) -> SqlHarnessRegistry:
    """Inject a SqlHarnessRegistry wired to the current session and tenant."""
    l5_url = os.environ.get("L5_BASE_URL")
    if l5_url and os.environ.get("L5_SERVICE_TOKEN"):
        return await make_live_l5_registry(
            session=db,
            tenant_id=_ctx.tenant_id,
        )
    return await make_sql_registry(session=db)


HarnessRegistryDep = Annotated[SqlHarnessRegistry, Depends(get_harness_registry)]
AuthCtxDep = Annotated[RequestContext, Depends(require_authenticated)]
# Gate decisions require content_admin or higher (reviewer/admin role).
ContentAdminCtxDep = Annotated[RequestContext, Depends(require_content_admin)]


# ---------------------------------------------------------------------------
# Run management
# ---------------------------------------------------------------------------


@router.post(
    "/harness/runs",
    response_model=RunResponse,
    status_code=201,
    summary="Create a harness run",
)
async def create_run(
    body: CreateRunRequest,
    registry: HarnessRegistryDep,
    ctx: AuthCtxDep,
) -> RunResponse:
    """Create a new HarnessRun for the authenticated tenant."""

    run = registry.create_run(
        tenant_id=ctx.tenant_id,
        workflow_type=body.workflow_type,
        initiated_by=body.initiated_by,
        account_id=body.account_id,
        value_pack_id=body.value_pack_id,
    )
    return RunResponse.from_domain(run)


@router.get(
    "/harness/runs",
    response_model=RunListResponse,
    summary="List harness runs",
)
async def list_runs(
    registry: HarnessRegistryDep,
    ctx: AuthCtxDep,
    status: HarnessRunStatus | None = Query(None),
    workflow_type: HarnessWorkflowType | None = Query(None),
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
) -> RunListResponse:
    """List harness runs for the authenticated tenant."""
    runs, total = await registry.list_runs(
        tenant_id=ctx.tenant_id,
        status=status,
        workflow_type=workflow_type,
        limit=limit,
        offset=offset,
    )
    return RunListResponse(
        items=[RunResponse.from_domain(r) for r in runs],
        total=total,
        limit=limit,
        offset=offset,
        has_more=(offset + len(runs)) < total,
    )


@router.get(
    "/harness/runs/{run_id}",
    response_model=RunResponse,
    summary="Get a harness run",
)
async def get_run(
    run_id: str,
    registry: HarnessRegistryDep,
    ctx: AuthCtxDep,
) -> RunResponse:
    """Get a single harness run by ID."""
    try:
        run = registry.get_run(run_id, ctx.tenant_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run {run_id} not found")
    except HarnessRegistryError as exc:
        raise HTTPException(status_code=403, detail=str(exc))
    return RunResponse.from_domain(run)


@router.post(
    "/harness/runs/{run_id}/transition",
    response_model=TransitionResponse,
    summary="Advance run state",
)
async def transition_run(
    run_id: str,
    body: TransitionRequest,
    registry: HarnessRegistryDep,
    ctx: AuthCtxDep,
) -> TransitionResponse:
    """Advance the state machine for a harness run."""
    try:
        run, event = registry.transition(
            run_id=run_id,
            tenant_id=ctx.tenant_id,
            to_state=body.to_state,
            validation_results=body.validation_results,
            human_override=body.human_override,
            state_payload=body.state_payload,
        )
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run {run_id} not found")
    except HarnessRegistryError as exc:
        raise HTTPException(status_code=403, detail=str(exc))
    except Exception as exc:
        logger.exception("Unexpected error transitioning run %s", run_id)
        raise HTTPException(status_code=400, detail=str(exc))

    return TransitionResponse(
        run=RunResponse.from_domain(run),
        trace_event=TraceEventResponse.from_domain(event) if event else None,
    )


@router.delete(
    "/harness/runs/{run_id}",
    status_code=204,
    summary="Cancel a harness run",
)
async def cancel_run(
    run_id: str,
    registry: HarnessRegistryDep,
    ctx: AuthCtxDep,
) -> None:
    """Cancel (transition to CANCELLED) a harness run."""
    try:
        run = registry.get_run(run_id, ctx.tenant_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run {run_id} not found")
    except HarnessRegistryError as exc:
        raise HTTPException(status_code=403, detail=str(exc))

    if run.current_state.is_terminal:
        raise HTTPException(
            status_code=400,
            detail=f"Run {run_id} is already in terminal state {run.current_state}",
        )

    try:
        registry.transition(
            run_id=run_id,
            tenant_id=ctx.tenant_id,
            to_state=HarnessState.CANCELLED,
            human_override=True,
        )
    except Exception as exc:
        logger.exception("Error cancelling run %s", run_id)
        raise HTTPException(status_code=400, detail=str(exc))


# ---------------------------------------------------------------------------
# Checkpoints
# ---------------------------------------------------------------------------


@router.get(
    "/harness/runs/{run_id}/checkpoints",
    response_model=CheckpointListResponse,
    summary="List checkpoints for a run",
)
async def list_checkpoints(
    run_id: str,
    registry: HarnessRegistryDep,
    ctx: AuthCtxDep,
) -> CheckpointListResponse:
    """List all checkpoints for a harness run."""
    try:
        registry.get_run(run_id, ctx.tenant_id)  # tenant check
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run {run_id} not found")
    except HarnessRegistryError as exc:
        raise HTTPException(status_code=403, detail=str(exc))

    checkpoints = registry.get_checkpoints(run_id, ctx.tenant_id)
    return CheckpointListResponse(
        items=[CheckpointResponse.from_domain(c) for c in checkpoints],
        total=len(checkpoints),
    )


@router.get(
    "/harness/runs/{run_id}/checkpoints/latest",
    response_model=CheckpointResponse,
    summary="Get latest checkpoint for a run",
)
async def get_latest_checkpoint(
    run_id: str,
    registry: HarnessRegistryDep,
    ctx: AuthCtxDep,
) -> CheckpointResponse:
    """Get the most recent checkpoint for a harness run."""
    try:
        registry.get_run(run_id, ctx.tenant_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run {run_id} not found")
    except HarnessRegistryError as exc:
        raise HTTPException(status_code=403, detail=str(exc))

    checkpoints = registry.get_checkpoints(run_id, ctx.tenant_id)
    if not checkpoints:
        raise HTTPException(status_code=404, detail=f"No checkpoints found for run {run_id}")
    latest = max(checkpoints, key=lambda c: c.created_at)
    return CheckpointResponse.from_domain(latest)


# ---------------------------------------------------------------------------
# Human gates
# ---------------------------------------------------------------------------


@router.get(
    "/harness/runs/{run_id}/gates",
    response_model=GateListResponse,
    summary="List gates for a run",
)
async def list_gates(
    run_id: str,
    registry: HarnessRegistryDep,
    ctx: AuthCtxDep,
) -> GateListResponse:
    """List all human gates for a harness run."""
    try:
        registry.get_run(run_id, ctx.tenant_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run {run_id} not found")
    except HarnessRegistryError as exc:
        raise HTTPException(status_code=403, detail=str(exc))

    gates = registry.list_gates_for_run(run_id, ctx.tenant_id)
    return GateListResponse(
        items=[GateResponse.from_domain(g) for g in gates],
        total=len(gates),
    )


@router.post(
    "/harness/runs/{run_id}/gates",
    response_model=GateResponse,
    status_code=201,
    summary="Create a human gate",
)
async def create_gate(
    run_id: str,
    body: CreateGateRequest,
    registry: HarnessRegistryDep,
    ctx: AuthCtxDep,
) -> GateResponse:
    """Create a human gate for a harness run."""
    try:
        registry.get_run(run_id, ctx.tenant_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run {run_id} not found")
    except HarnessRegistryError as exc:
        raise HTTPException(status_code=403, detail=str(exc))

    gate = registry.create_human_gate(
        run_id=run_id,
        tenant_id=ctx.tenant_id,
        gate_type=body.gate_type,
    )
    return GateResponse.from_domain(gate)


@router.post(
    "/harness/gates/{gate_id}/decide",
    response_model=GateResponse,
    summary="Decide a human gate",
)
async def decide_gate(
    gate_id: str,
    body: GateDecisionRequest,
    registry: HarnessRegistryDep,
    ctx: ContentAdminCtxDep,
) -> GateResponse:
    """Approve, reject, modify, or expire a human gate.

    Requires content_admin, tenant_admin, or super_admin role.
    decision_by is always server-derived from ctx.user_id — any
    body-supplied value is ignored to prevent spoofing.
    """
    try:
        _gate = registry.get_gate(gate_id, ctx.tenant_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Gate {gate_id} not found")
    except HarnessRegistryError as exc:
        raise HTTPException(status_code=403, detail=str(exc))

    decision_map = {
        "approved": GateStatus.APPROVED,
        "rejected": GateStatus.REJECTED,
        "modified": GateStatus.MODIFIED,
        "expired": GateStatus.EXPIRED,
    }

    # decision_by is always server-derived — never from request body.
    server_decision_by = ctx.user_id or ctx.tenant_id

    try:
        updated_gate = registry.decide_gate(
            gate_id=gate_id,
            tenant_id=ctx.tenant_id,
            new_status=decision_map[body.decision],
            decision_by=server_decision_by,
            decision_reason=body.decision_reason,
        )
    except (ValueError, HarnessRegistryError) as exc:
        raise HTTPException(status_code=400, detail=str(exc))

    return GateResponse.from_domain(updated_gate)


# ---------------------------------------------------------------------------
# Validation
# ---------------------------------------------------------------------------


@router.post(
    "/harness/runs/{run_id}/validate",
    response_model=ValidateClaimsResponse,
    summary="Validate claims for a run",
)
async def validate_claims(
    run_id: str,
    body: ValidateClaimsRequest,
    registry: HarnessRegistryDep,
    ctx: AuthCtxDep,
) -> ValidateClaimsResponse:
    """Validate a list of claims through the L5 ValidationHook."""
    try:
        registry.get_run(run_id, ctx.tenant_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run {run_id} not found")
    except HarnessRegistryError as exc:
        raise HTTPException(status_code=403, detail=str(exc))

    domain_requests = [
        DomainClaimValidationRequest(
            tenant_id=ctx.tenant_id,
            claim_id=c.claim_id,
            claim_text=c.claim_text,
            evidence_refs=c.evidence_refs,
            value_pack_id=c.value_pack_id,
            account_id=c.account_id,
        )
        for c in body.claims
    ]

    try:
        results = await registry.validate_claims(ctx.tenant_id, domain_requests)
    except HarnessRegistryError as exc:
        raise HTTPException(status_code=400, detail=str(exc))

    from ...harness.models import ValidationState

    result_responses = [ValidationResultResponse.from_domain(r) for r in results]
    return ValidateClaimsResponse(
        results=result_responses,
        total=len(results),
        passed=sum(1 for r in results if r.validation_state == ValidationState.PASSED),
        failed=sum(1 for r in results if r.validation_state == ValidationState.FAILED),
        needs_review=sum(1 for r in results if r.validation_state == ValidationState.NEEDS_REVIEW),
        insufficient_evidence=sum(
            1 for r in results if r.validation_state == ValidationState.INSUFFICIENT_EVIDENCE
        ),
    )


# ---------------------------------------------------------------------------
# Health
# ---------------------------------------------------------------------------


@router.get(
    "/harness/health",
    response_model=HarnessHealthResponse,
    summary="Harness health check",
)
async def harness_health(
    registry: HarnessRegistryDep,
) -> HarnessHealthResponse:
    """Return harness health: validation availability and L5 connectivity."""
    validation_available = registry.validation_available

    l5_healthy = False
    if validation_available:
        try:
            l5_healthy = await registry._validation.is_available()
        except Exception:
            l5_healthy = False

    # DB health: if we got this far the session is alive
    db_healthy = True

    status = "ok" if (db_healthy and (not validation_available or l5_healthy)) else "degraded"

    return HarnessHealthResponse(
        status=status,
        validation_available=validation_available,
        l5_healthy=l5_healthy,
        db_healthy=db_healthy,
    )
